numpy/core/src/umath/matmul.c.src

```diff
@@ -0,0 +1,365 @@
+{
+    npy_intp dOuter = *dimensions++;
+    npy_intp iOuter;
+    npy_intp s0 = *steps++;
+    npy_intp s1 = *steps++;
+    npy_intp s2 = *steps++;
+    npy_intp dm = dimensions[0];
+    npy_intp dn = dimensions[1];
+    npy_intp dp = dimensions[2];
+    npy_intp is1_m=steps[0], is1_n=steps[1], is2_n=steps[2], is2_p=steps[3],
+         os_m=steps[4], os_p=steps[5];
+    npy_intp ib1_n, ib2_n, ib2_p, ob_p;
+#if @USEBLAS@ & defined(HAVE_CBLAS)
+    npy_bool special_case = (dm == 1 || dn == 1 || dp == 1);
+    npy_bool scalar_out = (dm ==1 && dp == 1);
+    npy_bool scalar_vec = (dn == 1 && (dp == 1 || dm == 1));
+    npy_bool too_big_for_blas = (dm > NPY_MAX_INT || dn > NPY_MAX_INT ||
+                                 dp >= NPY_MAX_INT);
+    npy_bool input_contiguous = ((is1_m == sizeof(@typ@) ||
+                                  is1_n == sizeof(@typ@)) &&
+                                 (is2_n == sizeof(@typ@) ||
+                                  is2_p == sizeof(@typ@)));
+    npy_bool vector_matrix = ((dm == 1) &&
+                   (is2_n == sizeof(@typ@) || (is2_n == sizeof(@typ@) * dp)));
+    npy_bool matrix_vector = ((dp == 1)  &&
+                   (is1_n == sizeof(@typ@) || (is1_n == sizeof(@typ@) * dm)));
+#endif
+
+    ib1_n = is1_n*dn;
+    ib2_n = is2_n*dn;
+    ib2_p = is2_p*dp;
+    ob_p  = os_p *dp;
+
+    if (dn == 0) {
+        /* No operand, need to zero the output */
+        for (iOuter = 0; iOuter < dOuter; iOuter++,
+                        args[0] += s0, args[1] += s1, args[2] += s2) {
+            npy_intp m, p;
+            char *op=args[2];
+            for (m = 0; m < dm; m++) {
+                for (p = 0; p < dp; p++) {
+#if @SPECL@ == 1
+                    (*(@typ@ *)op).real = 0;
+                    (*(@typ@ *)op).imag = 0;
+#else
+                    *(@typ@ *)op = 0;
+#endif
+                    op  +=  os_p;
+                }
+                op  +=  os_m - ob_p;
+            }
+        }
+        return;
+    }
+    for (iOuter = 0; iOuter < dOuter; iOuter++,
+                         args[0] += s0, args[1] += s1, args[2] += s2) {
+        char *ip1=args[0], *ip2=args[1], *op=args[2];
+#if @USEBLAS@ & defined(HAVE_CBLAS)
+        /*
+         * TODO: refactor this out to a inner_loop_selector, in
+         * PyUFunc_MatmulLoopSelector. But that call does not have access to
+         * n, m, p and strides.
+         */
+        if (too_big_for_blas) {
+            @TYPE@_matmul_inner_noblas(ip1, ip2, op, dm, dn, dp,
+                      ib1_n, ib2_n, ib2_p, ob_p,
+                      is1_m, is1_n, is2_n, is2_p, os_m, os_p);
+        }
+        else if (special_case) {
+            /* Special case variants that have a 1 in the core dimensions */
+            if (scalar_out) {
+                /* row @ column, 1,1 output */
+                @TYPE@_dot(ip1, is1_n, ip2, is2_n, op, dn, NULL);
+            } else if (scalar_vec){
+                /*
+                 * 0d @ vector or vector @ 0d
```

Comment:
`0d` is misleading - `arr0d @ x` is an error, so never reaches this branch.

I think here you mean `length-1` - this branch is hit for `(1,1) @ x` or `(1,) @ x`

Reply:
Fixed the comment